src/GraphQL/Validation/IVariableVisitorProvider.cs

```diff
@@ -9,6 +9,13 @@ public interface IVariableVisitorProvider
         /// <summary>
         /// Gets a visitor for the specified validation context.
         /// </summary>
-        IVariableVisitor GetVisitor(ValidationContext context);
+        IVariableVisitor? GetVisitor(ValidationContext context);
+
+        /// <summary>
+        /// Prepares and returns a node visitor to be used to validate a document (via a node walker) against this
+        /// validation rule after the document has parsed all arguments. Validation failures are added then by this
+        /// visitor to a list stored within <see cref="ValidationContext.Errors"/>.
+        /// </summary>
+        ValueTask<INodeVisitor?> ValidateArgumentsAsync(ValidationContext context);
```

Comment:
The method name is a bit confusing. The existing `GetVisitor` fits well for the `IVariableVisitorProvider`, but `ValidateArgumentsAsync` looks inappropriate. I don't have a better suggestion, though. Maybe it should be placed in a new interface? Also, it's called `Validate` but returns visitor, which is not intuitive at all.

Reply:
What about if we add a whole new interface for validation rules?  We can then re-think and consolidate the entire design without breaking compatibility with rules designed for v7.

Reply:
> What about if we add a whole new interface for validation rules?

Maybe. I need to dive into the validation code to understand how it's implemented today.

Reply:
Sure.  I'd appreciate it if you could help me pick the best approach.  FYI this is what happens now (effectively):

1. `DocumentExecuter.ExecuteAsync(ExecutionOptions options)` runs with two relevant options:
  a. `IEnumerable<IValidationRule>? ValidationRules`
  b. `IEnumerable<IValidationRule>? CachedDocumentValidationRules`
2. If `ValidationRules` is null, then `ValidationRules = CoreRules` which contains the default set of rules
3. Adding rules via `AddValidationRule<T>(bool useForCachedDocuments = false)` will append to the proper property
4. If using document caching, and a cache hit occurs, `ValidationRules = CachedDocumentValidationRules`
5. `CachedDocumentValidationRules` is not used after this point
6. The document is parsed (if not using a cached instance)
7. Each instance of `IValidationRule.ValidateAsync` is executed.  Any non-null returned `INodeVisitor` instances are added to a list.  It is possible for validation to occur here and not use a node visitor.  But typically this method just returns a node visitor, often a new instance to hold temporary information during the validation.
8. Assuming there are any `INodeVisitor` instances, then `TypeInfo` is prepended to the list and the node visitor executes.  Each node is walked, and for each node, all the configured node visitors run for that node.  So the validation rules all run 'concurrently' rather than one after another.  The `TypeInfo` populates `ValidationContext` with information that assists validation rules to run efficiently.
9. If there are any errors, they are returned and execution stops
10. For each `IValidationRule`, `if (rule is IVariableVisitorProvider)` then add returned `IVariableVisitor` to a list of variable visitors (if not null)
11. Input values are parsed (and `Parser` and `Validator` delegates run if configured).  Any configured variable visitors are called at the appropriate time.  Unfortunately this only occurs with variables and not literals.  Ideally, I would like to change this to support both ... or deprecate this feature entirely in lieu of `Parser` and `Validator` delegates.
12. If any errors have occurred, they are returned and execution stops
13. NEW FEATURE:
  a. For each `IValidationRule`, `if (rule is IVariableVisitorProvider)`, call `ValidateArgumentsAsync` and add the returned `INodeVisitor` to a list (if not null)
  b. If any visitors, `TypeInfo` is prepended and these visitors execute in the same fashion as the first pass of validation rules.  The only difference is that it occurs after input value parsing has occurred.
14. If any errors have occurred, they are returned and execution stops
15. Document is executed

Reply:
Ok I’ll come up with an alternative solution and present it for review